pkg/servercounter/lease_counter_test.go

```diff
@@ -0,0 +1,249 @@
+	panic("not implemented")
+}
+
+func TestServerLeaseCounter(t *testing.T) {
+	testCases := []struct {
+		name string
+
+		templates        []leaseTemplate
+		leaseListerError error
+
+		labelSelector string
+		fallbackCount int
+
+		want int
+	}{
+		{
+			name:          "returns fallback count when no leases exist",
+			templates:     []leaseTemplate{},
+			labelSelector: "label=value",
+			fallbackCount: 999,
+			want:          999,
+		}, {
+			name: "returns fallback count when no leases matching selector exist",
+			templates: []leaseTemplate{
+				{
+					durationSecs:     1000,
+					timeSinceAcquire: time.Second,
+					labels:           labelMap{"label": "wrong_value"},
+				},
+				{
+					durationSecs:     1000,
+					timeSinceAcquire: time.Second,
+					labels:           labelMap{"wrong_label": "value"},
+				},
+			},
+			labelSelector: "label=value",
+			fallbackCount: 999,
+			want:          999,
+		}, {
+			name: "returns fallback count when no leases matching selector are still valid",
+			templates: []leaseTemplate{
+				{
+					durationSecs:     1000,
+					timeSinceAcquire: 10000 * time.Second,
+					labels:           labelMap{"label": "value"},
+				},
+				{
+					durationSecs:     1000,
+					timeSinceAcquire: time.Second,
+					labels:           labelMap{"wrong_label": "value"},
+				},
+			},
+			labelSelector: "label=value",
+			fallbackCount: 999,
+			want:          999,
+		}, {
+			name:             "returns fallbackCount when LeaseLister returns an error",
+			templates:        []leaseTemplate{},
+			labelSelector:    "label=value",
+			fallbackCount:    999,
+			leaseListerError: fmt.Errorf("test error"),
+			want:             999,
+		}, {
+			name: "counts only valid leases matching label selector",
+			templates: []leaseTemplate{
+				{
+					durationSecs:     1000,
+					timeSinceAcquire: time.Second,
+					labels:           labelMap{"label": "value"},
+				},
+				{
+					durationSecs:     1000,
+					timeSinceAcquire: time.Second,
+					labels:           labelMap{"label": "value"},
+				},
+				{
+					durationSecs:     1000,
+					timeSinceAcquire: time.Second,
+					labels:           labelMap{"label": "wrong_value"},
+				},
+			},
+			labelSelector: "label=value",
+			fallbackCount: 999,
+			want:          2,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			leases := make([]*coordinationv1.Lease, len(tc.templates))
+			for i, template := range tc.templates {
+				leases[i] = newLeaseFromTemplate(template)
+			}
+			lister := &fakeLeaseLister{
+				leases: leases,
+				err:    tc.leaseListerError,
+			}
+
+			counter, err := NewServerLeaseCounter(lister, tc.labelSelector, tc.fallbackCount)
+			if err != nil {
+				t.Fatalf("server counter creation failed: %v", err)
+			}
+
+			got := counter.CountServers()
+			if tc.want != got {
+				t.Errorf("incorrect server count (got: %v, want: %v)", got, tc.want)
+			}
+		})
+	}
+}
```

Comment:
can you add a test that ensures the fallback value is updated on successful List, and the updated fallback count is used when the next List returns an error?

Reply:
Done.